### PR TITLE
RPED can carry & install computer circuit boards and complete computer frames & general code cleanup

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
Read Tin

- If you have multiple computer circuit boards in the RPED a tgui input is displayed asking you which specific one you want to install in the computer frame
- If you have cable & glass inside the RPED it will attempt to install them in the computer frame as well

You still have to use a screwdriver to turn the computer on as usual

Added extra checks to ensure we only install/remove a circuit board when we actually a circuit board to work with

## Why It's Good For The Game
RPED already has the ability to carry cable & glass sheets which is required to complete a computer frame, now after adding the ability to carry & install computer circuit boards we can build computers much faster without the need for manual clicking.

## Changelog
:cl:
add: RPED can carry & install computer boards and complete computer frames
/:cl:

### CODE REVIEW

**Comment by SyncIt21** [comment]
> Does this apply to the bluespace RPED as well?

Yes

**Comment by TheVekter** [comment]
> > Does this apply to the bluespace RPED as well?
> 
> Yes

This is interesting. I'm wondering if this will have balance implications? I can't think of any off the top of my head, though, so maybe I'm just being paranoid.

**Comment by ExcessiveUseOfCobblestone** [comment]
> This is interesting. I'm wondering if this will have balance implications? I can't think of any off the top of my head, though, so maybe I'm just being paranoid.

paranoid, youd need TK to start worry about a wonky interaction.

only thing I can think of would be a bug that actually replaced the board while an existing computer exists, but I dont see anything here that would create the scenario.

**Comment by YuliceTheDruid** [comment]
How will this affect engi borgs (I hate having to ask about silicons)

**Comment by SyncIt21** [comment]
> How will this affect engi borgs (I hate having to ask about silicons)

Same benifits

**Comment by zxaber** [comment]
>How will this affect engi borgs

Engiborgs have an upgrade on the same research tier that carries a single circuit board for machines and computers. There's no real balance adjustment here. But I was very proud of the circuit apparatus, so this makes me sad that it's basically the same benefit but better.

**Comment by optimumtact** [comment]
I don't think I want this without a cohesive construction overhaul such as the RAT tbh.

**Comment by SyncIt21** [comment]
> I don't think I want this without a cohesive construction overhaul such as the RAT tbh.

The RPED can carry a maximum of 30 glass sheets & cable coil so even though you can carry a large number of circuit boards you can only install a maximum of 6 computers[each one takes 5 cable & 2 glass sheets] before you have to refill the RPED for cable.

Compare that to the duffle/construction bag which can carry a maximum of 12 circuit boards & 30 glass sheets + 60 cable[2 stacks] the RPED is pretty insignificant & underwhelming. It's either that or people should carry a separate bag just for cable & glass if they ever want to build a large number of computers

And then we have additional storage cost of carrying around iron sheets to make the actual computer frames the maximum number of computers we can actually build are very much limited by all these factors So you know the construction implications aren't that bad

This is really convenient, hopefully you will consider this.

**Comment by optimumtact** [comment]
I'm sure someone will

**Comment by SyncIt21** [comment]
This had been sitting for a while. @Time-Green maybe you are interested in this feature :) ?

**Comment by Time-Green** [comment]
> This had been sitting for a while. @Time-Green maybe you are interested in this feature :) ?

I'm sorry Sync, I don't feel strongly enough about this. I don't really know any maintainers/headcoders very into construction (except Oranges' design doc)

**Comment by SyncIt21** [comment]
> > This had been sitting for a while. @Time-Green maybe you are interested in this feature :) ?
> 
> I'm sorry Sync, I don't feel strongly enough about this. I don't really know any maintainers/headcoders very into construction (except Oranges' design doc)

Thats fair. On the flip side he and no one else has any strong opinions against this, also no balance implications from the above mentioned comments so no one would have problems if this got merged and is beneficial for engi players like me so...? It's fine all upto you friend.

**Comment by SyncIt21** [comment]
It's been a month :( . This is a really sweet feature, I need somebody to give it just a little attention.

@jlsnow301 since you upvoted this PR maybe you would like to merge/review this :) ? pretty please?

File: code/game/machinery/computer/buildandrepair.dm [hunk 1]
```diff
@@ -39,25 +54,51 @@
...
+					//attack again so we can install the cable & glass
+					attackby(replacer, user, params)
+					return
+
+			if(!circuit && istype(P, /obj/item/circuitboard/computer))
+				install_board(P, user, TRUE)
 				return
 
-			else if(istype(P, /obj/item/circuitboard) && !circuit)
+			else if(!circuit && istype(P, /obj/item/circuitboard))
```
Comment (Time-Green) [inline code/game/machinery/computer/buildandrepair.dm:91]: This is the third !circuit check in a row, could you branch it out?

File: code/game/machinery/computer/buildandrepair.dm [hunk 2]
```diff
@@ -73,17 +114,33 @@
...
+				var/obj/item/stack/cable_coil/cable = null
+				if(istype(P, /obj/item/stack/cable_coil))
+					cable = P
+				else if(istype(P, /obj/item/storage/part_replacer))
+					cable = locate(/obj/item/stack/cable_coil) in P.contents
+				if(!cable)
+					return
+
+				//install cable
+				if(!cable.tool_start_check(user, amount=5))
```
Comment (Time-Green) [inline code/game/machinery/computer/buildandrepair.dm:128]:
```suggestion
				if(!cable.tool_start_check(user, amount = 5))
```

File: code/game/machinery/computer/buildandrepair.dm [hunk 3]
```diff
@@ -73,17 +114,33 @@
...
 					return
 				to_chat(user, span_notice("You start adding cables to the frame..."))
-				if(P.use_tool(src, user, 20, volume=50, amount=5))
+				if(cable.use_tool(src, user, istype(P, /obj/item/storage/part_replacer) ? 0 : 20, volume=50, amount=5))
```
Comment (Time-Green) [inline code/game/machinery/computer/buildandrepair.dm:131]:
```suggestion
				if(cable.use_tool(src, user, istype(P, /obj/item/storage/part_replacer) ? 0 : 20, volume = 50, amount = 5))
```

File: code/game/machinery/computer/buildandrepair.dm [hunk 4]
```diff
@@ -94,19 +151,31 @@
 				var/obj/item/stack/cable_coil/A = new (drop_location(), 5)
 				if (!QDELETED(A))
 					A.add_fingerprint(user)
-				return
+			else
+				//serach for glass sheets which can either be the attacking item or inside an rped
```
Comment (Time-Green) [inline code/game/machinery/computer/buildandrepair.dm:155]:
```suggestion
				//search for glass sheets which can either be the attacking item or inside an rped
```

File: code/game/machinery/computer/buildandrepair.dm [hunk 5]
```diff
@@ -94,19 +151,31 @@
...
+					glass_sheets = P
+				else if(istype(P, /obj/item/storage/part_replacer))
+					glass_sheets = locate(/obj/item/stack/sheet/glass) in P.contents
+				if(!glass_sheets)
+					return
 
-			if(istype(P, /obj/item/stack/sheet/glass))
-				if(!P.tool_start_check(user, amount=2))
+				//install glass sheets
+				if(!glass_sheets.tool_start_check(user, amount=2))
```
Comment (Time-Green) [inline code/game/machinery/computer/buildandrepair.dm:165]:
```suggestion
				if(!glass_sheets.tool_start_check(user, amount = 2))
```

File: code/game/machinery/computer/buildandrepair.dm [hunk 6]
```diff
@@ -94,19 +151,31 @@
...
 					return
 				playsound(src, 'sound/items/deconstruct.ogg', 50, TRUE)
 				to_chat(user, span_notice("You start to put in the glass panel..."))
-				if(P.use_tool(src, user, 20, amount=2))
+				if(glass_sheets.use_tool(src, user, istype(P, /obj/item/storage/part_replacer) ? 0 : 20, amount=2))
```
Comment (Time-Green) [inline code/game/machinery/computer/buildandrepair.dm:169]:
```suggestion
				if(glass_sheets.use_tool(src, user, istype(P, /obj/item/storage/part_replacer) ? 0 : 20, amount = 2))
```

File: code/game/machinery/constructable_frame.dm [hunk 7]
```diff
@@ -208,12 +208,12 @@
...
 					return
 
-			if(istype(P, /obj/item/circuitboard/machine))
+			if(!circuit && istype(P, /obj/item/circuitboard/machine))
 				var/obj/item/circuitboard/machine/machine_board = P
 				install_board(machine_board, user, TRUE)
 				return
 
-			else if(istype(P, /obj/item/circuitboard))
+			else if(!circuit && istype(P, /obj/item/circuitboard))
```
Comment (Time-Green) [inline code/game/machinery/constructable_frame.dm:216]: Ditto on the !circuit checks

File: code/modules/research/stock_parts.dm [hunk 8]
```diff
@@ -19,19 +19,17 @@ If you create T5+ please take a pass at mech_fabricator.dm. The parts being good
 	create_storage(storage_type = /datum/storage/rped)
 
 // check to see if this rped have atleast one circuitboard
-/obj/item/storage/part_replacer/proc/has_an_circuitboard()
-	for(var/obj/item/circuitboard/machine/board in contents)
-		return TRUE
-	return FALSE
+/obj/item/storage/part_replacer/proc/has_an_circuitboard(board_type)
```
Comment (Time-Green) [inline code/modules/research/stock_parts.dm:22]: anbelievable

File: code/modules/research/stock_parts.dm [hunk 9]
```diff
@@ -19,19 +19,17 @@ If you create T5+ please take a pass at mech_fabricator.dm. The parts being good
...
+	return !isnull((locate(board_type) in contents))
```
Comment (Time-Green) [inline code/modules/research/stock_parts.dm:23]:
```suggestion
	return (locate(board_type) in contents)
```

I don't uhhhhhh... What?

**Review by Time-Green — CHANGES_REQUESTED** [review]
I have looked deep within myself, and I do like the feature. I don't think anyone is currently working on the construction rework, so I don't see it fair to neglect our current system yet. 

Man construction code has a lot of istype checks

File: code/modules/research/stock_parts.dm (hunk 9 above)
Comment (Time-Green) [reply code/modules/research/stock_parts.dm:23]: Oh wait nvm I get it, you're boolerizing the return value

**Comment by SyncIt21** [comment]
@Time-Green any further changes?

**Comment by Time-Green** [comment]
> @Time-Green any further changes?

You're free

**Comment by SyncIt21** [comment]
> > @Time-Green any further changes?
> 
> You're free

:)

**Pull request merged**